### PR TITLE
types(remix-server-runtime): make `context` mandatory in `DataFunctionArgs`

### PR DESCRIPTION
**Pull request by justinwaite**
Closes: #3988

My strategy for addressing this issue is to always pass an empty object for the context, rather than passing nothing (or undefined). Since we've already declared the AppLoadContext interface to be `{[key: string]: unknown}`, I feel like this would be a reasonable approach. 

- [ ] Docs
- [ ] Tests

Testing Strategy:

I created an app in the playground and (manually had to) copy over the built remix-server-runtime directory into it. I then augmented the AppLoadContext interface with a property called `notes`. I then tried to access both a known and unknown property to make sure that typescript knew the difference.

```typescript
declare module '@remix-run/server-runtime' {
  export interface AppLoadContext {
    notes: string[]
  }
}

export const loader: LoaderFunction = async ({ request, context }) => {
  console.log(context.asdf.toString());
  
  console.log(context.notes.map(n => n.toLowerCase())); // Look Ma! No undefined check for context!

  let userId = await requireUserId(request);
  let noteListItems = await getNoteListItems({ userId });
  return json<LoaderData>({ noteListItems });
};
```

Here you can see the difference. Since `asdf` isn't part of the AppLoadContext augmented interface, Typescript correctly infers it as `unknown`. However, since `notes` _was_ added to the interface, it correctly infers it as a `string[]`. Note in either case that we didn't need to add an undefined check for the context, since it will always be at least an empty object (`{}`).
Screen Shot 2022-08-12 at 6 17 05 PM
Screen Shot 2022-08-12 at 6 17 30 PM

### CODE REVIEW

File: packages/remix-server-runtime/server.ts [hunk 1]
```diff
@@ -110,7 +110,7 @@ async function handleDataRequest({
       match = getRequestMatch(url, matches);
 
       response = await callRouteAction({
-        loadContext,
+        loadContext: loadContext ?? {},
```
Comment (marwan38) [inline packages/remix-server-runtime/server.ts:113]: I would default this as high up as possible. So, in the args definition  line 79) or even higher up. I'm assuming that the consumer of this function doesn't really or shouldn't be instantiating the empty object.

**Review by pcattori — CHANGES_REQUESTED** [review]
Looks like this is on the right track!

Could you make all user-facing APIs (i.e. things exported from the main `index.ts`) be optional (`context?: AppLoadContext`) while making all the internal APIs required (`context: AppLoadContext`)?

For the user-facing optional ones, I agree with @marwan38 that you could use a default value in the function signature to simplify. For example, within the `createRequestHandler` function:

```ts
  return async function requestHandler(request, loadContext = {}) {
```

^I _think_ that might be the only place you'll need the default value of `= {}`.

**Comment by justinwaite** [comment]
> Looks like this is on the right track!
> 
> We'd like to make all user-facing APIs (i.e. things exported from the main `index.ts`) optional (`context?: AppLoadContext`) while making all the internal APIs required (`context: AppLoadContext`).
> 
> The only user-facing API should be the `requestHandler` function within the `createRequestHandler` function. I think all that's needed is to change that to:
> 
> ```ts
>   return async function requestHandler(request, loadContext = {}) {
> ```
> 
> For all other places, you should be able to make the load context required.

Hey @pcattori, what exactly would you like me to change here? The interface for `RequestHandler` remains the same:
```typescript
export type RequestHandler = (
  request: Request,
  loadContext?: AppLoadContext
) => Promise<Response>;
```

The only place I'm making it _required_ is in `callRouteAction` and `callRouteLoader`. The other changes I'm just using some nullish coalescing to ensure that _something_ is passed in, i.e. an empty object, in the event that the user passes null or undefined into the `requestHandler` for the `context`.

File: packages/remix-server-runtime/server.ts (hunk 1 above)
Comment (justinwaite) [reply packages/remix-server-runtime/server.ts]: Good call. I moved it up to the `requestHandler` to be the default value for the loadContext arg. This definitely did clean it up

**Review by pcattori — APPROVED** [review]

**Review by MichaelDeBoey — APPROVED** [review]

**Comment by pcattori** [comment]
> Hey @pcattori, what exactly would you like me to change here?

I think @marwan38 said _what_ should change more clearly than I did 😅 was just trying to emphasize that we want that change not only because its tidier, but because it correctly models that users should be able to omit `loadContext` arg and `requestHandler` is the only place users would do that.

**Comment by pcattori** [comment]
Waiting for #4010 to fix the Windows CI errors

**Comment by marwan38** [comment]
Nice! Thank you

**Pull request merged**